ngraph/frontend/onnx_import/src/editor/detail/subgraph_extraction.cpp

```diff
@@ -0,0 +1,405 @@
+    {
+        ++old_input_name;
+    }
+
+    m_node_inputs.erase(old_input_name);
+    m_node_inputs.insert({new_edge.m_node_idx, new_edge.m_tensor_name});
+}
+
+void SubgraphExtractor::extract_subgraph(std::vector<OutputEdge> subgraph_outputs)
+{
+    if (subgraph_outputs.empty())
+    {
+        subgraph_outputs = all_output_edges();
+    }
+
+    SubgraphComponents subgraph;
+
+    for (const auto& output_edge : subgraph_outputs)
+    {
+        subgraph += discover_output_contributors(output_edge, subgraph);
+    }
+
+    extract_subgraph_from_onnx_model(subgraph);
+}
+
+SubgraphExtractor::SubgraphComponents SubgraphExtractor::discover_output_contributors(
+    const OutputEdge& output_edge, const SubgraphComponents& already_collected) const
+{
+    const auto already_visited = [&already_collected](const int node_index) {
+        return already_collected.nodes.count(node_index) > 0;
+    };
+
+    SubgraphComponents output_contributors;
+    output_contributors.outputs.insert(output_edge.m_tensor_name);
+
+    // reverse DFS graph traversal
+    std::stack<int> nodes_to_visit;
+    nodes_to_visit.push(output_edge.m_node_idx);
+
+    while (!nodes_to_visit.empty())
+    {
+        const auto n = nodes_to_visit.top();
+        nodes_to_visit.pop();
+
+        if (already_visited(n))
+        {
+            continue;
+        }
+
+        output_contributors.nodes.insert(n);
+
+        // check if the visitor reached any of the graph inputs
+        // and/or keep looking for more contributors further up in the graph
+        const auto n_inputs = m_node_inputs.equal_range(n);
+        for (auto input_name = n_inputs.first; input_name != n_inputs.second; ++input_name)
+        {
+            if (is_graph_input(m_onnx_graph, input_name->second))
+            {
+                output_contributors.inputs.insert(input_name->second);
+                // when an initializer has a matching graph input
+                if (is_graph_initializer(m_onnx_graph, input_name->second))
+                {
+                    output_contributors.initializers.insert(input_name->second);
+                }
+            }
+            else if (is_graph_initializer(m_onnx_graph, input_name->second))
+            {
+                // when an initializer doesn't have a corresponding input
+                output_contributors.initializers.insert(input_name->second);
+            }
+            else
+            {
+                nodes_to_visit.push(find_source_node_idx(m_onnx_graph, n, input_name->second));
+            }
+        }
+    }
+
+    return output_contributors;
+}
+
+void SubgraphExtractor::extract_subgraph_from_onnx_model(const SubgraphComponents& subgraph)
+{
+    discard_by_name(*(m_onnx_graph.mutable_input()), subgraph.inputs);
+    discard_by_name(*(m_onnx_graph.mutable_initializer()), subgraph.initializers);
+    discard_by_name(*(m_onnx_graph.mutable_output()), subgraph.outputs);
```

Comment:
According to `discard_by_name` description its not meant to be used with outputs.

Reply:
Good catch, fixed the docs.